c10/cuda/CUDAStream.cpp

```diff
@@ -369,6 +382,35 @@ void setCurrentCUDAStream(CUDAStream stream) {
   current_streams[ptr->device_index] = ptr;
 }
 
+// registers an external stream to the custom pool
+CUDAStream registerCustomCUDAStream(DeviceIndex device_index, cudaStream_t stream) {
+  initCUDAStreamsOnce();
+
+  // Make sure the device is valid
+  cudaDeviceProp prop;
+  
+  auto device_query_result = cudaGetDeviceProperties(&prop, device_index);
+  TORCH_CHECK(device_query_result == cudaSuccess);
+
+  // Make sure the stream is good before we increase the counter
+  uint32_t flags;
+  auto query_result = cudaStreamGetFlags(stream, &flags);
+  TORCH_CHECK(query_result == cudaSuccess);
+
+  uint32_t stream_id = custom_counters[device_index]++;
+  // make sure the pool is not overflowing
+  TORCH_CHECK(stream_id < kStreamsPerPool);
```

Comment:
Sooo.... you do know that this means you're only going to get 32 custom streams? If you want to support more, you're going to have to do some more restructuring: you'll need to expand the amount of space for the stream id index at

```
// -- 25 bits -- -- 2 bits --  -- 5 bits -----
// zeros         StreamIdType  stream id index
```

and probably also have a dynamically allocated vector rather than a static array. This is related to another limitation with this current patch which is that you have no way of deregistering custom streams; once you've registered 32 that's it, you're done, no more. Maybe this is good enough for your use case?

Reply:
We isolate things into a bunch processes. Number of streams in a given process is usually only one.